File: May-LeetCoding-Challenge/16-Odd-Even-Linked-List/Odd-Even-Linked-List.java
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if(head==null){return head;}
        if(head.next==null){return head;}
        ListNode p = head;
        ListNode q = head.next;
        ListNode paire = head.next;
        ListNode cur = q.next;
        int i = 2;
        while(cur!=null){
            i++;
            if((i&1)==0){
                q.next = cur;
                q = q.next;
            }
            else{
                p.next = cur;
                p = p.next;
            }
            cur = cur.next;
        }
        q.next = null;
        p.next = paire;
        return head;// 1 2 4 5
    }
}

class Solution2 {
    public ListNode oddEvenList(ListNode head) {
        // we separate the list to odd list and event list, and we combine these two list
        if(head == null)    return null;
        
        ListNode odd = head;
        ListNode even = head.next;
        ListNode evenHead = even;
             
        while(even != null && even.next != null) {
            odd.next = even.next;
            odd = odd.next;
            even.next = odd.next;
            even = even.next;
        }
        
        odd.next = evenHead;
        return head;   
    }
}

File: May-LeetCoding-Challenge/16-Odd-Even-Linked-List/Odd-Even-Linked-List.scala
// Since the question asks for a O(1) space complexity and O(nodes) time complexity
// I decide to make no difference to a java solution...
object Solution {
    def oddEvenList(head: ListNode): ListNode = {
        if (head == null)
            null
        else{
            // Initialize two LinkedList
            var currentOdd = head
            var currentEven = head.next
            var firstEven = currentEven

            while (currentEven != null && currentEven.next != null) {
                currentOdd.next = currentEven.next
                currentOdd = currentOdd.next
                currentEven.next = currentOdd.next
                currentEven = currentEven.next
            }
            // Combine two LinkedList
            currentOdd.next = firstEven

            head
        }

    }
}

//nothing special ... but i HATE "null" in scala's code !!!
object Solution2 {

  def oddEvenList(head: ListNode): ListNode = {
    if (head == null || head.next == null) head
    else {
      val evenListHead = head.next
      def reLinkList(odd: ListNode, even: ListNode): Unit = {
        if (even == null || even.next == null) odd.next = evenListHead
        else {
          odd.next = even.next
          even.next = odd.next.next
          reLinkList(odd.next, even.next)
        }
      }
      reLinkList(head, evenListHead)
      head
    }
  }
}
